builder.md

```diff
@@ -1,57 +1,180 @@
+executed, and MUST look like `<output-dir>/base16-<slug><extension>`, where the
+`SLUG` is taken from the scheme filename made lowercase with spaces replaced
+with dashes and both the extension and `output-dir` are taken from
+`config.yaml`.
+
+If the build fails for any reasons, the program MUST exit with code `1` and MAY
+output an error message. This is not exhaustive and new exit codes might be
+added in the future to account for specific errors, so scripters catching
+generic errors SHOULD check for non-zero status.
+
+Otherwise, the program MUST exit with code `0` and an OPTIONAL success message.
+
+#### 3.1.3 CLI example:
+
+```bash
+git clone https://github.com/base16-project/base16-vim.git
+git clone https://github.com/base16-project/base16-schemes.git
+base16 base16-vim/templates base16-schemes/*.yaml
+```
+
+Should create a `colors` (on the current directory) containing `base16-XXX.vim`
+files.
+
+### 3.2. Library
+
+The other option is exposing a software library other developers may use to
+assist developing more complex base16-compatible tooling.
+
+As above, the library MUST implement single feature: building templates.
+
+This exposed library, or any internal code, has no specific required structure.
+
+The author MAY choose how (through a class, function that operates on file
+paths, etc) they will expose the builder functionality to the caller, according
+to their judgement.
+
+They SHOULD try to achieve an ergonomical and powerful interface and follow the
+best practices on their respective programming languages.
+
+Official reference implementations will follow semantic versioning matching
+this spec, in the format: `<base16-version>-builder_revision`.
 
-## Code Structure
-There is no outline for a recommended code structure that a base16 theme builder should follow but a design goal should be to keep the application as simple as possible providing only the functionality described in this document. If you feel you have a great idea for additional functionality please raise an issue in the [base16 repository](https://github.com/chriskempson/base16).
+## 4. Considerations
+Mustache was chosen as the templating language due to its simplicity and
+widespread adoption across languages. YAML was chosen to describe scheme and
+configuration files for the same reasons.
 
-## Considerations
-Mustache was chosen as the templating language due to its simplicity and widespread adoption across languages. YAML was chosen to describe scheme and configuration files for the same reasons.
+The core scheme repository was based off the single scheme repository so
+builders supporting v0.8-v0.9 of the spec can continue to function without
+changes.
 
-The core scheme repository was based off the single scheme repository so builders supporting v0.8-v0.9 of the spec can continue to function without changes.
+The revised builder functionality specification introduced in v0.11 did not
```

Comment:
Personally, I wouldn't include this line. Maybe it would make sense to move the changelog from the README into this file?

Additionally, maybe it's worthwhile putting our unwritten rule into writing that "spec upgrades should be backwards compatible for templates".

Reply:
Should I remove the `0.11` line but keep the `0.8` and `0.9` one?